File: app/src/main/java/com/rollapp/shared/data/repository/FirebaseAuthRepository.kt
package com.rollapp.shared.data.repository

import com.google.firebase.auth.EmailAuthProvider
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.auth.GoogleAuthProvider
import com.google.firebase.auth.UserProfileChangeRequest
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.SetOptions
import com.rollapp.shared.core.AppError
import com.rollapp.shared.core.FirestorePaths
import com.rollapp.shared.core.Outcome
import com.rollapp.shared.core.firebaseCall
import com.rollapp.shared.domain.model.User
import com.rollapp.shared.domain.repository.AuthRepository
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.tasks.await

@Singleton
class FirebaseAuthRepository @Inject constructor(
    private val auth: FirebaseAuth,
    private val firestore: FirebaseFirestore
) : AuthRepository {

    override val currentUser: Flow<User?> = callbackFlow {
        val listener = FirebaseAuth.AuthStateListener { a ->
            trySend(a.currentUser?.toDomain())
        }
        auth.addAuthStateListener(listener)
        awaitClose { auth.removeAuthStateListener(listener) }
    }

    override fun currentUid(): String? = auth.currentUser?.uid

    override suspend fun signInWithGoogle(idToken: String): Outcome<User> = firebaseCall {
        val credential = GoogleAuthProvider.getCredential(idToken, null)
        val result = auth.signInWithCredential(credential).await()
        val user = result.user ?: throw IllegalStateException("Sign-in returned no user")
        upsertProfile(user).also { profile ->
            // A Google account always has a name; anonymous upgrades may not.
            if (profile.name.isBlank()) updateAuthDisplayName(user, "Someone")
        }
    }

    override suspend fun signInWithEmail(email: String, password: String): Outcome<User> {
        val validation = validateEmailPassword(email, password)
        if (validation != null) return Outcome.Failure(validation)
        return firebaseCall {
            val result = auth.signInWithEmailAndPassword(email.trim(), password).await()
            val user = result.user ?: throw IllegalStateException("Sign-in returned no user")
            upsertProfile(user)
        }
    }

    override suspend fun signUpWithEmail(name: String, email: String, password: String): Outcome<User> {
        if (name.isBlank()) return Outcome.Failure(AppError.Validation("What should we call you?"))
        val validation = validateEmailPassword(email, password)
        if (validation != null) return Outcome.Failure(validation)
        return firebaseCall {
            val result = auth.createUserWithEmailAndPassword(email.trim(), password).await()
            val user = result.user ?: throw IllegalStateException("Sign-up returned no user")
            updateAuthDisplayName(user, name.trim())
            upsertProfile(user, overrideName = name.trim())
        }
    }

    override suspend fun signInAnonymously(): Outcome<User> = firebaseCall {
        val result = auth.signInAnonymously().await()
        val user = result.user ?: throw IllegalStateException("Anonymous sign-in returned no user")
        upsertProfile(user, overrideName = "Guest")
    }

    override suspend fun linkAnonymousToEmail(name: String, email: String, password: String): Outcome<User> {
        val current = auth.currentUser
            ?: return Outcome.Failure(AppError.NotAuthenticated)
        if (!current.isAnonymous) return Outcome.Failure(AppError.Validation("This account is already permanent"))
        val validation = validateEmailPassword(email, password)
        if (validation != null) return Outcome.Failure(validation)

        return firebaseCall {
            val credential = EmailAuthProvider.getCredential(email.trim(), password)
            val result = current.linkWithCredential(credential).await()
            val user = result.user ?: current
            updateAuthDisplayName(user, name.trim())
            // The uid is preserved by linking, so every existing membership survives.
            upsertProfile(user, overrideName = name.trim(), markPermanent = true)
        }
    }

    override suspend fun linkAnonymousToGoogle(idToken: String): Outcome<User> {
        val current = auth.currentUser ?: return Outcome.Failure(AppError.NotAuthenticated)
        if (!current.isAnonymous) return Outcome.Failure(AppError.Validation("This account is already permanent"))
        return firebaseCall {
            val credential = GoogleAuthProvider.getCredential(idToken, null)
            val result = current.linkWithCredential(credential).await()
            val user = result.user ?: current
            upsertProfile(user, markPermanent = true)
        }
    }

    override suspend fun sendPasswordReset(email: String): Outcome<Unit> {
        if (!email.contains("@")) return Outcome.Failure(AppError.Validation("Enter a valid email"))
        return firebaseCall { auth.sendPasswordResetEmail(email.trim()).await() }
    }

    override suspend fun signOut(): Outcome<Unit> = firebaseCall {
        auth.signOut()
    }

    override suspend fun deleteAccount(): Outcome<Unit> {
        val user = auth.currentUser ?: return Outcome.Failure(AppError.NotAuthenticated)
        return firebaseCall {
            // The profile doc goes first: once the auth user is gone the client can no
            // longer satisfy the `request.auth.uid == userId` rule that guards it.
            firestore.collection(FirestorePaths.USERS).document(user.uid).delete().await()
            user.delete().await()
        }
    }

    /**
     * Creates `users/{uid}` on first sign-in and refreshes the mutable fields on every
     * later one. `merge` keeps fields the client does not own — notification prefs,
     * device tokens — intact.
     */
    private suspend fun upsertProfile(
        user: FirebaseUser,
        overrideName: String? = null,
        markPermanent: Boolean = false
    ): User {
        val doc = firestore.collection(FirestorePaths.USERS).document(user.uid)
        val name = overrideName
            ?: user.displayName?.takeIf { it.isNotBlank() }
            ?: user.email?.substringBefore("@")
            ?: "Guest"

        val data = mutableMapOf<String, Any?>(
            "name" to name,
            "email" to (user.email ?: ""),
            "photoUrl" to user.photoUrl?.toString(),
            "isAnonymous" to (user.isAnonymous && !markPermanent),
            "updatedAt" to FieldValue.serverTimestamp()
        )

        val existing = doc.get().await()
        if (!existing.exists()) {
            data["createdAt"] = FieldValue.serverTimestamp()
        } else if (existing.getString("photoUrl") != null && user.photoUrl == null) {
            // Don't clear a photo the user uploaded themselves just because the
            // auth provider has none.
            data.remove("photoUrl")
        }

        doc.set(data, SetOptions.merge()).await()

        return User(
            uid = user.uid,
            name = name,
            email = user.email.orEmpty(),
            photoUrl = user.photoUrl?.toString() ?: existing.getString("photoUrl"),
            createdAt = System.currentTimeMillis(),
            isAnonymous = user.isAnonymous && !markPermanent
        )
    }

    private suspend fun updateAuthDisplayName(user: FirebaseUser, name: String) {
        user.updateProfile(
            UserProfileChangeRequest.Builder().setDisplayName(name).build()
        ).await()
    }

    private fun validateEmailPassword(email: String, password: String): AppError? = when {
        email.isBlank() || !email.contains("@") -> AppError.Validation("Enter a valid email")
        password.length < 6 -> AppError.WeakPassword()
        else -> null
    }

    private fun FirebaseUser.toDomain() = User(
        uid = uid,
        name = displayName?.takeIf { it.isNotBlank() } ?: email?.substringBefore("@") ?: "Guest",
        email = email.orEmpty(),
        photoUrl = photoUrl?.toString(),
        createdAt = metadata?.creationTimestamp ?: 0L,
        isAnonymous = isAnonymous
    )
}

File: ios/WeWere/Data/Repository/FirebaseAuthRepository.swift
import Foundation
import Combine
import FirebaseAuth
import FirebaseFirestore

final class FirebaseAuthRepository: AuthRepository {

    private let auth: Auth
    private let firestore: Firestore

    init(auth: Auth, firestore: Firestore) {
        self.auth = auth
        self.firestore = firestore
        currentUser = auth.userStream().map { $0.map(Self.toDomain) }.eraseToAnyPublisher()
    }

    let currentUser: Stream<User?>

    func currentUid() -> String? { auth.currentUser?.uid }

    func signInWithGoogle(idToken: String, accessToken: String) async -> Outcome<User> {
        await firebaseCall {
            let credential = GoogleAuthProvider.credential(withIDToken: idToken, accessToken: accessToken)
            let result = try await auth.signIn(with: credential)
            let profile = try await provisionProfile(result.user)
            // A Google account always has a name; anonymous upgrades may not.
            if profile.name.trimmingCharacters(in: .whitespaces).isEmpty {
                try await updateAuthDisplayName(result.user, "Someone")
            }
            return profile
        }
    }

    func signInWithEmail(email: String, password: String) async -> Outcome<User> {
        if let validation = validateEmailPassword(email, password) { return .failure(validation) }
        return await firebaseCall {
            let result = try await auth.signIn(withEmail: email.trimmingCharacters(in: .whitespaces), password: password)
            return try await provisionProfile(result.user)
        }
    }

    func signUpWithEmail(name: String, email: String, password: String) async -> Outcome<User> {
        let trimmedName = name.trimmingCharacters(in: .whitespaces)
        if trimmedName.isEmpty { return .failure(.validation("What should we call you?")) }
        if let validation = validateEmailPassword(email, password) { return .failure(validation) }
        return await firebaseCall {
            let result = try await auth.createUser(withEmail: email.trimmingCharacters(in: .whitespaces), password: password)
            try await updateAuthDisplayName(result.user, trimmedName)
            return try await provisionProfile(result.user, overrideName: trimmedName)
        }
    }

    func signInAnonymously() async -> Outcome<User> {
        await firebaseCall {
            let result = try await auth.signInAnonymously()
            return try await provisionProfile(result.user, overrideName: "Guest")
        }
    }

    func linkAnonymousToEmail(name: String, email: String, password: String) async -> Outcome<User> {
        guard let current = auth.currentUser else { return .failure(.notAuthenticated) }
        if !current.isAnonymous { return .failure(.validation("This account is already permanent")) }
        if let validation = validateEmailPassword(email, password) { return .failure(validation) }
        let trimmedName = name.trimmingCharacters(in: .whitespaces)

        return await firebaseCall {
            let credential = EmailAuthProvider.credential(withEmail: email.trimmingCharacters(in: .whitespaces), password: password)
            let result = try await current.link(with: credential)
            let user = result.user
            try await updateAuthDisplayName(user, trimmedName)
            // The uid is preserved by linking, so every existing membership survives.
            return try await provisionProfile(user, overrideName: trimmedName, markPermanent: true)
        }
    }

    func linkAnonymousToGoogle(idToken: String, accessToken: String) async -> Outcome<User> {
        guard let current = auth.currentUser else { return .failure(.notAuthenticated) }
        if !current.isAnonymous { return .failure(.validation("This account is already permanent")) }
        return await firebaseCall {
            let credential = GoogleAuthProvider.credential(withIDToken: idToken, accessToken: accessToken)
            let result = try await current.link(with: credential)
            return try await provisionProfile(result.user, markPermanent: true)
        }
    }

    func sendPasswordReset(email: String) async -> Outcome<Void> {
        if !email.contains("@") { return .failure(.validation("Enter a valid email")) }
        return await firebaseCall {
            try await auth.sendPasswordReset(withEmail: email.trimmingCharacters(in: .whitespaces))
        }
    }

    func signOut() async -> Outcome<Void> {
        await firebaseCall { try auth.signOut() }
    }

    func deleteAccount() async -> Outcome<Void> {
        guard let user = auth.currentUser else { return .failure(.notAuthenticated) }
        return await firebaseCall {
            // The profile doc goes first: once the auth user is gone the client can no
            // longer satisfy the `request.auth.uid == userId` rule that guards it.
            try await firestore.collection(FirestorePaths.users).document(user.uid).delete()
            try await user.delete()
        }
    }

    /// Makes sure the signed-in user has a profile document, repairing accounts whose
    /// first attempt never landed. Idempotent and safe to call on every launch.
    func ensureProfile() async -> Outcome<Void> {
        guard let user = auth.currentUser else { return .failure(.notAuthenticated) }
        return await firebaseCall {
            let doc = try await firestore.collection(FirestorePaths.users).document(user.uid).getDocument()
            if !doc.exists {
                _ = try await provisionProfile(user, overrideName: user.isAnonymous ? "Guest" : nil)
            }
        }
    }

    /// Runs the profile write in a detached task and waits for it.
    ///
    /// The caller is a ViewModel whose screen is about to go away: signing in flips
    /// the auth-state listener, which navigates away from the sign-in screen. Handing
    /// the work to a task that nobody cancels means a cancellation stops the
    /// *waiting*, not the write.
    private func provisionProfile(_ user: FirebaseAuth.User, overrideName: String? = nil,
                                  markPermanent: Bool = false) async throws -> User {
        let firestore = self.firestore
        return try await Task.detached(priority: .userInitiated) {
            try await Self.upsertProfile(firestore: firestore, user: user, overrideName: overrideName, markPermanent: markPermanent)
        }.value
    }

    /// Creates `users/{uid}` on first sign-in and refreshes the mutable fields on every
    /// later one. `merge` keeps fields the client does not own — notification prefs,
    /// device tokens — intact.
    private static func upsertProfile(firestore: Firestore, user: FirebaseAuth.User,
                                      overrideName: String?, markPermanent: Bool) async throws -> User {
        let doc = firestore.collection(FirestorePaths.users).document(user.uid)
        let name = overrideName ?? Self.fallbackName(for: user)

        var data: [String: Any] = [
            "name": name,
            "email": user.email ?? "",
            "photoUrl": user.photoURL?.absoluteString as Any,
            "isAnonymous": user.isAnonymous && !markPermanent,
            "updatedAt": FieldValue.serverTimestamp()
        ]

        let existing = try await doc.getDocument()
        if !existing.exists {
            data["createdAt"] = FieldValue.serverTimestamp()
        } else if existing.str("photoUrl") != nil && user.photoURL == nil {
            // Don't clear a photo the user uploaded themselves just because the
            // auth provider has none.
            data.removeValue(forKey: "photoUrl")
        }

        try await doc.setData(data, merge: true)

        return User(
            uid: user.uid,
            name: name,
            email: user.email ?? "",
            photoUrl: user.photoURL?.absoluteString ?? existing.str("photoUrl"),
            createdAt: Date.nowMillis,
            isAnonymous: user.isAnonymous && !markPermanent
        )
    }

    private func updateAuthDisplayName(_ user: FirebaseAuth.User, _ name: String) async throws {
        let change = user.createProfileChangeRequest()
        change.displayName = name
        try await change.commitChanges()
    }

    private func validateEmailPassword(_ email: String, _ password: String) -> AppError? {
        let trimmed = email.trimmingCharacters(in: .whitespaces)
        if trimmed.isEmpty || !trimmed.contains("@") { return .validation("Enter a valid email") }
        if password.count < 6 { return .weakPassword() }
        return nil
    }

    /// Display name, else the part of the email before the @, else "Guest".
    private static func fallbackName(for user: FirebaseAuth.User) -> String {
        if let display = user.displayName, !display.trimmingCharacters(in: .whitespaces).isEmpty { return display }
        if let email = user.email, let local = email.split(separator: "@").first { return String(local) }
        return "Guest"
    }

    private static func toDomain(_ user: FirebaseAuth.User) -> User {
        User(
            uid: user.uid,
            name: fallbackName(for: user),
            email: user.email ?? "",
            photoUrl: user.photoURL?.absoluteString,
            createdAt: user.metadata.creationDate?.millis ?? 0,
            isAnonymous: user.isAnonymous
        )
    }
}
